src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java

```diff
@@ -901,43 +1006,94 @@ public void reduceNeededResources(final Block block)
+            return EntityList.createEntityFromNBT(entityInfo.entityData, world);
+        }
+        catch (Exception e)
+        {
+            Log.getLogger().info("Couldn't restore entitiy", e);
+            return null;
+        }
+    }
 
-                entityHanging.posX += pos.getX();
-                entityHanging.posY += pos.getY();
-                entityHanging.posZ += pos.getZ();
-                //also sets position based on tile
-                entityHanging.setPosition(
-                        entityHanging.getHangingPosition().getX(),
-                        entityHanging.getHangingPosition().getY(),
-                        entityHanging.getHangingPosition().getZ());
+    private Boolean spawnEntity(@NotNull final Structure.StructureBlock currentBlock)
+    {
+        final Template.EntityInfo entityInfo = currentBlock.entity;
+        if (entityInfo != null && job instanceof JobBuilder)
+        {
+            final Entity entity = getEntityFromEntityInfoOrNull(entityInfo);
 
-                entityHanging.setWorld(world);
-                entityHanging.dimension = world.provider.getDimension();
+            final PlacementSettings settings = ((JobBuilder) job).getStructure().structure().getStructure().getSettings();
 
-                world.spawnEntityInWorld(entityHanging);
-            }
-            else if (entity instanceof EntityMinecart)
+            final BlockPos pos = currentBlock.blockPosition;
+            if (entity != null)
             {
-                @Nullable final EntityMinecart minecart = (EntityMinecart) entity;
-                minecart.posX += pos.getX();
-                minecart.posY += pos.getY();
-                minecart.posZ += pos.getZ();
+                final List<ItemStack> request = new ArrayList<>();
 
-                minecart.setWorld(world);
-                minecart.dimension = world.provider.getDimension();
+                if(entity instanceof EntityItemFrame)
+                {
+                    final ItemStack stack = ((EntityItemFrame) entity).getDisplayedItem();
+                    stack.stackSize = 1;
+                    request.add(stack);
+                    request.add(new ItemStack(Items.ITEM_FRAME, 1));
+                }
+                else if(entity instanceof EntityArmorStand)
+                {
+                    request.add(entity.getPickedResult(new RayTraceResult(worker)));
+                    entity.getArmorInventoryList().forEach(request::add);
+                }
+                else
+                {
+                    request.add(entity.getPickedResult(new RayTraceResult(worker)));
+                }
 
-                world.spawnEntityInWorld(minecart);
+                for(final ItemStack stack: request)
+                {
+                    if (checkOrRequestItems(stack))
+                    {
+                        return false;
+                    }
+                }
+
+                for(final ItemStack stack: request)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Combine this loop with the one that starts on line 1059. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS3047)